### PR TITLE
Handle last event before exit

### PR DESCRIPTION
**Pull request by framillien**
Always send last event (if any) before handling exit.

### CODE REVIEW

**Comment by santigimeno** [comment]
I think you're right in the case of `pcsclite.cpp` but not in `cardreader.cpp`. Notice that in https://github.com/santigimeno/node-pcsclite/blob/master/src/cardreader.cpp#L397 we send the last event.

One improvement for `cardreader.cpp` would be that in case of exit condition, don't perform 2 consecutive `uv_async_send` (https://github.com/santigimeno/node-pcsclite/blob/master/src/cardreader.cpp#L397 and https://github.com/santigimeno/node-pcsclite/blob/master/src/cardreader.cpp#L402) but just one, and then implement the modification you've proposed

**Comment by framillien** [comment]
Ok, going down threads, events, ... :smile: I discover lib_uv, but I have some interrogations:
- You always send the same async callback struct. lib_uv copy it on send method ? Or use the same ? If it's the same, we could be in the first handler function with the exit flag alreay set, right ?
- If I look at the doc: http://docs.libuv.org/en/v1.x/async.html There is a warning about deduplicated events. Can we fall in this case when the two events are fired ?

**Comment by santigimeno** [comment]
@framillien OMG i didn't know about that warning. There is a problem then. We'll have to synchronize the threads to avoid calling `uv_async_send` consecutive times without having handled the callback. I'll also look into it. If we synchronize the threads correctly there should be no issue with using the same async struct.

**Comment by santigimeno** [comment]
Thanks for pointing that out

**Comment by framillien** [comment]
Hum ... I don't like locks ! Structure update must be securized, ok, but do not try to force sent of all callbacks, I think. We can loose some of them, important thing is to have the current state of readers. For me, this mean:
- Structure update/read must be secure and you already have a mutex on reader, sounds good.
- Last event (error, end, ...) must be handled.
- Best effort for others (if it's fired good, but don't use complex locking for this)

So, as you say, I think we can handle exit flag like you do it in pcsclite. I complete my patch with this.

**Comment by santigimeno** [comment]
@framillien I'm not to sure about that. I don't think the order is guaranteed by `uv_async_send` so maybe we don't received the last state of the readers

**Comment by santigimeno** [comment]
Hmm, maybe I'm wrong. By reading the libuv code I think it could be guaranteed. Not sure though :(

**Comment by santigimeno** [comment]
Yes I think it is, as the communication between threads is done via a pipe

**Comment by framillien** [comment]
Ok, I have pushed two new commits.
In this version, the last status was sent, and on windows it's an error event. So I skip it, if it's related to card reader removed (like in pcsclite loop).
In my tests cases, I never miss usefull events.

File: src/cardreader.cpp [hunk 1]
```diff
@@ -392,20 +393,24 @@ void CardReader::HandlerFunction(void* arg) {
 
         if (!keep_watching) {
             reader->m_state = 2;
+            if (result == SCARD_E_NO_READERS_AVAILABLE) { // It's not an error, we send reader removed event (do_exit==true)
```
Comment (santigimeno) [inline src/cardreader.cpp]: We can move this to the `HandleReaderStatusChange` function in the same way we do in `pcsclite.cpp`

**Comment by santigimeno** [comment]
@framillien Apart from that comment it LGTM. I'll try to test it before merging it. Could you please squash it into one commit? Thanks!

Regarding the missing events. I'm not sure just getting the last one is always the best option. As an example: the case that a reader is disconnected and connected immediately and that the application misses the disconnected event. It could be that the application is designed to do something if the reader has been disconnected regardless the reader is connected again. The application won't be able to do that. I think the best option is to give the application every event pcscd emits so it's up to the application how to proceed. What do you think?

In any case, your patches makes things better.

**Comment by framillien** [comment]
Well, we are talking about thread synchronisation. If we start talking about human interaction, yes, I suppose that our event was handled by lib uv before the user can think about plugging his card again :wink:. Events we can eventually loose are flags like : card in use, ... For me, we can't do safe treatment based on this flags. But maybe someone could need it.

For now, I update the merge request.

**Comment by santigimeno** [comment]
LGTM. I'll try to merge it during the weekend. Thanks!!!

**Comment by framillien** [comment]
Thanks to you too.

**Comment by santigimeno** [comment]
@framillien I had merged your PR in version `0.4.3` (https://github.com/santigimeno/node-pcsclite/commit/3800c392918159fed582b2626591153725b29912) but had to revert it because it was causing crashes and unexpected error messages. One of the crashes happened when running the `example.js` with one reader with a card inserted. I think the problem lies in deciding when we should be emit the event before exiting the threads. For example, if the user cancels the checking for new readers via `pcsc.close()`, we should not be emitting the event but just exit. That's just one case, I'll look into it more thoroughly when I have time.

**Comment by framillien** [comment]
I see. Your test is on linux ? I also take a look begin of next week.

**Comment by santigimeno** [comment]
Yes. It fails both in Linux and OS X

**Comment by framillien** [comment]
Hello, I have commit a small update, but problem is still here, I try to check this week, sorry for delay.

**Comment by framillien** [comment]
Well, about our last events problems, I propose another option: based on m_state value.
From code I see that 'm_state' could be:
- 0 thread running
- 1 user requested exit
- 2 exit due to some errors

So on event handler we can do:
- 0 => Handle success and errors events (same as now)
- 1 => User request exit, so, do not throw any events else 'end' for cardreader
- 2 => Throw only 'error' event to inform user about current problem

I rework this branch like this ?

**Comment by santigimeno** [comment]
@framillien that sounds about right. Last weekend i tried that approach, and it *almost* worked. I didn't have much time, so I couldn't finish it. It would be great if you could. Thanks for taking the time to look into this

**Comment by framillien** [comment]
Ok, can you take a look and do some tests on this version ? I have done some tests with examples.js and our applications: It's look good for me.

I just have an interrogation about one exit condition (take a look in pcsclite.cpp), and sometime I receive an event twice when unplug a card reader (status 0xb), but I don't think it's a problem. 

Thanks,

File: src/cardreader.cpp [hunk 2]
```diff
@@ -379,33 +383,32 @@ void CardReader::HandlerFunction(void* arg) {
...
         uv_mutex_lock(&reader->m_mutex);
         if (reader->m_state == 1) {
+            // Exit requested by user. Notify close method about SCardStatusChange was interrupted.
             uv_cond_signal(&reader->m_cond);
-        }
-
-        if (!keep_watching) {
+            async_baton->async_result->do_exit = true;
+        } else if (result != (LONG)SCARD_S_SUCCESS) {
+            // Exit this loop due to errors
```
Comment (santigimeno) [inline src/cardreader.cpp]: You could do:
```
async_baton->async_result->do_exit = (reader->m_state != 0);
```
once after the ifs

File: src/pcsclite.cpp [hunk 3]
```diff
@@ -124,12 +124,10 @@ void PCSCLite::HandleReaderStatusChange(uv_async_t *handle, int status) {
     AsyncBaton* async_baton = static_cast<AsyncBaton*>(handle->data);
     AsyncResult* ar = async_baton->async_result;
 
-    if (ar->do_exit) {
-        uv_close(reinterpret_cast<uv_handle_t*>(&async_baton->async), CloseCallback); // necessary otherwise UV will block
-        return;
-    }
-
-    if ((ar->result == SCARD_S_SUCCESS) || (ar->result == (LONG)SCARD_E_NO_READERS_AVAILABLE)) {
+    if (async_baton->pcsclite->m_state==1) {
```
Comment (santigimeno) [inline src/pcsclite.cpp]: spaces around the `==`

File: src/cardreader.cpp [hunk 4]
```diff
@@ -333,6 +333,27 @@ void CardReader::HandleReaderStatusChange(uv_async_t *handle, int status) {
 
     AsyncResult* ar = async_baton->async_result;
 
+    if (reader->m_state==1) {
```
Comment (santigimeno) [inline src/cardreader.cpp]: spaces around the `==`

File: src/cardreader.cpp [hunk 5]
```diff
@@ -333,6 +333,27 @@ void CardReader::HandleReaderStatusChange(uv_async_t *handle, int status) {
...
+        // Swallow events : Listening thread was cancelled by user.
+    } else if ((ar->result == SCARD_S_SUCCESS)
```
Comment (santigimeno) [inline src/cardreader.cpp]: I prefer having the `||` at the end of the line

File: src/pcsclite.cpp [hunk 6]
```diff
@@ -124,12 +124,10 @@ void PCSCLite::HandleReaderStatusChange(uv_async_t *handle, int status) {
...
+        // Swallow events : Listening thread was cancelled by user.
+    } else if ((ar->result == SCARD_S_SUCCESS)
```
Comment (santigimeno) [inline src/pcsclite.cpp]: I prefer having the `||` at the end of the line

File: src/pcsclite.cpp [hunk 7]
```diff
@@ -180,17 +184,18 @@ void PCSCLite::HandlerFunction(void* arg) {
...
-            }
-
-            if (result != SCARD_S_SUCCESS) {
+            } else if (result != SCARD_S_SUCCESS) {
                 pcsclite->m_state = 2;
             }
-
             uv_mutex_unlock(&pcsclite->m_mutex);
         } else {
+            // Really need to exit if previous result is not success ?
```
Comment (santigimeno) [inline src/pcsclite.cpp]: I think we should return an error  and exit

**Comment by santigimeno** [comment]
@framillien It looks great. My preliminary tests have been successful. Left some comments on the code.
Thanks!

**Comment by framillien** [comment]
Ok, error and typo changed.

**Comment by santigimeno** [comment]
It looks good, but I'm getting some crashes from time to time due to some synchronization issues. Looking into it.
```
Disconnected
node: ../src/cardreader.cpp:73: virtual CardReader::~CardReader(): Assertion `ret == 0' failed.
Aborted

```

**Comment by santigimeno** [comment]
Got a fix for that

**Comment by santigimeno** [comment]
@framillien Could you give a try to https://github.com/santigimeno/node-pcsclite/tree/pr46? It has your changes with some fixes. Thanks!

**Comment by santigimeno** [comment]
@framillien Have you had time to test the branch?

**Comment by santigimeno** [comment]
@framillien  did you get to test it? Thanks!

**Comment by santigimeno** [comment]
@framillien I included your changes here: https://github.com/santigimeno/node-pcsclite/commit/06d31638e241467c061f3688f1697fcce24b410a. They have been published in `0.4.6`. Thanks!

**Pull request closed without merging**